kolibri/core/bookmarks/api.py

```diff
@@ -0,0 +1,74 @@
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.status import HTTP_200_OK
+from rest_framework.status import HTTP_201_CREATED
+from rest_framework.status import HTTP_204_NO_CONTENT
+from rest_framework.status import HTTP_400_BAD_REQUEST
+from rest_framework.status import HTTP_403_FORBIDDEN
+from rest_framework.status import HTTP_405_METHOD_NOT_ALLOWED
+
+from .models import Bookmark
+from .serializers import BookmarksSerializer
+from kolibri.core.api import ValuesViewset
+
+
+class BookmarksViewSet(ValuesViewset):
+    values = ("channel_id", "contentnode_id", "id", "content_id")
+    serializer_class = BookmarksSerializer
+    permission_classes = [IsAuthenticated]
+
+    def list(self, request):
+        bookmarks = Bookmark.objects.filter(facility_user=request.user).values(
+            "channel_id", "content_id", "contentnode_id", "id"
+        )
+        serializer = BookmarksSerializer(bookmarks, many=True)
+
+        return Response(serializer.data, status=HTTP_200_OK)
+
+    def create(self, request):
+        """
+        Overridden for two reasons:
+        1) Inject user into the data before creating it to avoid NOT NULL on facility_user
```

Comment:
You are reimplementing a lot of the logic for a serializer here - would be easier to overrride the `create` method of the serializer instead, and inject the user at that stage:
```
def create(self, validated_data):
    if "request" in self.context:
        user = self.context["request"].user
        validated_data["user"] = user
    try:
        return super(BookmarkSerializer, self).create(validated_data)
    except IntegrityError:
        return Bookmark.objects.get(**validated_data)
```

Alternatively, can just require that the user is submitted with the request, and lean on the permissions above to validate that the user creating the bookmark is allowed to do so (this then leaves open the possibility of an admin creating a bookmark for someone else).